src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,370 @@
+{
+    TxoGroup ret;
+    if (prev != nullptr)
+        ret = *prev;
+    else
+        ret.first = 0;
+    // Don't add an element that already exists  -- allow this, bad solutions eliminated later
+    // assert(ret.second.find(i) == ret.second.end());
+    ret.first += i->first; // update the amount
+    COutput outp = i->second;
+    // verify that the output value is consistent -- TODO: Check pcoin and make sure it doesn't get deleted -- located
+    // in mapWallet
+    assert(outp.tx->vout[outp.i].nValue == i->first);
+    ret.second.insert(i); // add this txo to the set
+    return ret;
+}
+
+// Take a group that is not a solution and find a bunch of solutions by appending additional TXOs onto it.
+// This is a recursive function that is limited by MAX_TXOs and MAX_ELECTIVE_TXOS
+bool extend(const CAmount targetValue,
+    /*const*/ SpendableTxos &available,
+    TxoGroup grp,
+    TxoGroupMap &solutions,
+    int depth)
+{
+    if (depth >= MAX_TXOS)
+        return false; // unambiguously break the recursion
+    SpendableTxos::iterator aEnd = available.end(); // For speed, get these once
+    SpendableTxos::iterator aBeg = available.begin();
+    bool ret = false;
+    SpendableTxos::iterator small;
+    small = available.lower_bound(targetValue - grp.first); // Find a value at or above what we need
+    if (1)
+    {
+        SpendableTxos::iterator i = small;
+        // while not at the end and there's too little value or we already used this output then keep going.
+        // while((i != aEnd) && ((i->first + grp.first < targetValue) || (grp.second.find(i) != grp.second.end()))  )
+        // allow dups and eliminate later for performance
+        while ((i != aEnd) && (i->first + grp.first < targetValue)) // iterate to make sure it is actually big enough.
+        {
+            ++i;
+        }
+
+        if (i != aEnd) // Ok we have a solution so add it.
+        {
+            TxoGroup g = makeGroup(i, &grp);
+            solutions.insert(TxoGroupMap::value_type(g.first, g));
+            // LogPrint("wallet","CoinSelection solution found: excess %ld, txos: %d\n", g.first - targetValue,
+            // g.second.size());
+            ret = true;
+        }
+    }
+    if (small == aEnd)
+        --small;
+    for (int i = 0; i < 5; i++) // check 5 more close values
+    {
+        // lower_bound returns an element >= the passed value, so get one that is smaller
+        while ((small != aBeg) && (small->first + grp.first > targetValue))
+            --small;
+        // keep looking if there are no solutions or if the depth is low
+        if ((solutions.empty() || (depth < MAX_ELECTIVE_TXOS)))
+        {
+            // Its better performance to accidentally add a double than to check every time
+            // go to the prev if this one is already in the group.
+            // while ((small != aBeg) && (grp.second.find(small) != grp.second.end())) --small;
+            // if (small != aBeg)
+            {
```

Comment:
looks like some commented out code , while/if , that maybe aren't needed.

Reply:
I want to leave this one in.  The comment above described the situation, which is that if this code is commented, it will occasionally provide an invalid solution (that is later detected and eliminated), because it can add the same TXI into the solution set twice.  So this comment describes a correctness/performance tradeoff that may change in the future.